src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorProjectChangePublisher.cs

```diff
@@ -209,9 +211,27 @@ protected virtual void DeleteFile(string publishFilePath)
 
         protected virtual void SerializeToFile(ProjectSnapshot projectSnapshot, string publishFilePath)
         {
+            // We need to avoid having an incomplete file at any point, but our project.razor.json is large enough that it will be written
+            // as multiple operations
             var fileInfo = new FileInfo(publishFilePath);
-            using var writer = fileInfo.CreateText();
-            _serializer.Serialize(writer, projectSnapshot);
+            var tempFilePath = string.Concat(publishFilePath, TempFileExt);
+            var tempFileInfo = new FileInfo(tempFilePath);
+
+            // This needs to be in explicit brackets because the operation needs to be completed
+            // by the time we move the tempfile into its place
+            using (var writer = tempFileInfo.CreateText())
+            {
+                _serializer.Serialize(writer, projectSnapshot);
+
+                if (fileInfo.Exists)
+                {
+                    var oldFilePath = string.Concat(publishFilePath, OldFileExt);
+                    fileInfo.MoveTo(oldFilePath);
```

Comment:
Why the rename to .old and then delete?

Also what happens if an existing .old or .tmp file exists? (imagine file operations throwing and interrupting all of the logic here)